### PR TITLE
Manage advanced name converters for documentation generation

### PR DESCRIPTION
**Pull request by Soullivaneuh**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #2343
| License       | MIT
| Doc PR        | N/A

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
I don't know how to fix PHPStan. The interface is unknow to it because Symfony 4.2 is not stable yet.

File: src/Hydra/Serializer/DocumentationNormalizer.php [hunk 1]
```diff
@@ -185,7 +186,9 @@ private function getHydraProperties(string $resourceClass, ResourceMetadata $res
             }
 
             if ($this->nameConverter) {
-                $propertyName = $this->nameConverter->normalize($propertyName);
+                $propertyName = $this->nameConverter instanceof AdvancedNameConverterInterface
```
Comment (dunglas) [inline src/Hydra/Serializer/DocumentationNormalizer.php]: You can drop the test, just pass all arguments: they will be ignored for implementations not implementing `AdvancedNameConverterInterface`.
Also, can you pass the the format and the context?

Same for the other one

**Comment by dunglas** [comment]
Just add the PHPStan errors to the ignore list, the are expected.

**Comment by Soullivaneuh** [comment]
@dunglas Requested changes done, except for some required parameter. I don't know how to fetch them properly here.

**Comment by dunglas** [comment]
For both classes, you'll have to propagate `$format` and `$context` from `normalize()` to the private method.

**Comment by Soullivaneuh** [comment]
@dunglas Changes made.

**Comment by Soullivaneuh** [comment]
@dunglas Is anything missing to get this merged ?

I can't access to `SymfonyInsight` informations.

**Comment by dunglas** [comment]
Basically, you'll have to add the new `$format` and `$context` arguments to all private methods, and propagate these variable from `normalize` to the name converter.

File: src/Hydra/Serializer/DocumentationNormalizer.php [hunk 2]
```diff
@@ -178,14 +178,15 @@ private function getPropertyNameCollectionFactoryContext(ResourceMetadata $resou
     private function getHydraProperties(string $resourceClass, ResourceMetadata $resourceMetadata, string $shortName, string $prefixedShortName): array
     {
         $properties = [];
-        foreach ($this->propertyNameCollectionFactory->create($resourceClass, $this->getPropertyNameCollectionFactoryContext($resourceMetadata)) as $propertyName) {
+        $context = $this->getPropertyNameCollectionFactoryContext($resourceMetadata);
```
Comment (Soullivaneuh) [inline src/Hydra/Serializer/DocumentationNormalizer.php]: > Sorry but it still doesn't look good to me. The context must be propagated from normalize and denormalize, not re-created.

Are you talking about this line?

May you help me a bit or edit my PR? I don't know how to manage the Hydra part for that. :-|

**Comment by Soullivaneuh** [comment]
> Basically, you'll have to add the new $format and $context arguments to all private methods,

The format is a private constant, why not using it directly? :thinking:

**Comment by dunglas** [comment]
Do you need any help on this @Soullivaneuh?

**Comment by Soullivaneuh** [comment]
> Do you need any help on this

help

:arrow_right: https://github.com/api-platform/core/pull/2344/files#r237442798

The `$serializerContext` is not created by me on the Swagger normalizer.

About Hydra, I don't know. I indeed use `$this->getPropertyNameCollectionFactoryContext($resourceMetadata)` but it was already done on the `foreach`.

What do you need exactly? :thinking:

**Comment by dunglas** [comment]
You must not recreate the context. You must forward the one passed to `normalize` and `denormalize`, that's all.

**Comment by Soullivaneuh** [comment]
@dunglas  ?

**Comment by dunglas** [comment]
You also need to upgrade the phpstan config (previous errors can now be ignored)

**Comment by Soullivaneuh** [comment]
> You also need to upgrade the phpstan config

I did a rebase, I think you already fixed that.

**Comment by Soullivaneuh** [comment]
PHPStan is not yelling anymore, let's wait other services.

@dunglas Are we clean about the code diff?

**Comment by dunglas** [comment]
Yes I think so. ping @api-platform/core-team

**Comment by dunglas** [comment]
Thanks @Soullivaneuh!

**Pull request merged**